### PR TITLE
consistency decoder

### PR DESCRIPTION
**Pull request by williamberman**
```py
from diffusers import StableDiffusionPipeline, ConsistencyDecoderVae
import torch

pipe = StableDiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16
)
pipe.to('cuda')
pipe("horse", generator=torch.Generator('cpu').manual_seed(0)).images[0].save('orig.png')

vae = ConsistencyDecoderVae.from_pretrained("../consistency-decoder")
pipe = StableDiffusionPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, vae=vae
)
pipe.to('cuda')

pipe("horse", generator=torch.Generator('cpu').manual_seed(0)).images[0].save('new.png')


```

orig
orig

new
new

```py
import torch
from diffusers import ConsistencyDecoderVae, AutoencoderKL
from diffusers.utils.loading_utils import load_image
import numpy as np
from PIL import Image

vae_orig = AutoencoderKL.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16, subfolder="vae")
vae_orig.to('cuda')

vae_new = ConsistencyDecoderVae.from_pretrained("../consistency-decoder", torch_dtype=torch.float16)
vae_new.to('cuda')


image = load_image("https://raw.githubusercontent.com/openai/consistencydecoder/main/assets/gt1.png").resize((256, 256))
image = torch.from_numpy(np.array(image).transpose(2, 0, 1).astype(np.float32) / 127.5 - 1)[None, :, :, :].half().cuda()
latent = vae_new.encode(image).latent_dist.mean

sample_orig = vae_orig.decode(latent).sample
sample_new = vae_new.decode(latent).sample

Image.fromarray(((sample_orig.detach()[0].cpu().numpy().transpose(1, 2, 0) + 1) * 127.5).clip(0, 255).astype(np.uint8)).save('orig.png')
Image.fromarray(((sample_new.detach()[0].cpu().numpy().transpose(1, 2, 0) + 1) * 127.5).clip(0, 255).astype(np.uint8)).save('new.png')

```

original
orig

new
new

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_consistency_decoder.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        num_inference_steps: Optional[int] = None,
+        device: Union[str, torch.device] = None,
+    ):
+        assert num_inference_steps == 2, "TODO"
+        self.timesteps = torch.tensor([1008, 512], dtype=torch.long, device=device)
+        self.sqrt_alphas_cumprod = self.sqrt_alphas_cumprod.to(device)
+        self.sqrt_one_minus_alphas_cumprod = self.sqrt_one_minus_alphas_cumprod.to(device)
+        self.c_skip = self.c_skip.to(device)
+        self.c_out = self.c_out.to(device)
+        self.c_in = self.c_in.to(device)
```
Comment (williamberman) [inline src/diffusers/schedulers/scheduling_consistency_decoder.py:73]: Need to implement more timestep options but I'm ok shipping w/out and adding a better error message for now

File: src/diffusers/models/unet_2d.py [hunk 2]
```diff
@@ -144,6 +145,9 @@ def __init__(
         elif time_embedding_type == "positional":
             self.time_proj = Timesteps(block_out_channels[0], flip_sin_to_cos, freq_shift)
             timestep_input_dim = block_out_channels[0]
+        elif time_embedding_type == "learned":
+            self.time_proj = nn.Embedding(num_train_timesteps, block_out_channels[0])
+            timestep_input_dim = block_out_channels[0]
```
Comment (sayakpaul) [inline src/diffusers/models/unet_2d.py:150]: Is this change relevant?

File: src/diffusers/pipelines/consistency_models/pipeline_consistency_decoder.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        output_type: Optional[str] = "pil",
+        num_inference_steps=2,
+        return_dict: bool = True,
+    ):
+        means = torch.tensor([0.38862467, 0.02253063, 0.07381133, -0.0171294], device=z.device, dtype=z.dtype)[
+            None, :, None, None
+        ]
+        stds = torch.tensor([0.9654121, 1.0440036, 0.76147926, 0.77022034], device=z.device, dtype=z.dtype)[
+            None, :, None, None
+        ]
```
Comment (sayakpaul) [inline src/diffusers/pipelines/consistency_models/pipeline_consistency_decoder.py:43]: Wonder if it's better off as a config.

**Comment by sayakpaul** [comment]
@williamberman summary of the changes:

* Docs
* Tests (including a slow one)

File: src/diffusers/schedulers/scheduling_consistency_decoder.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+
+class ConsistencyDecoderScheduler(SchedulerMixin, ConfigMixin):
+    order = 1
+
+    @register_to_config
+    def __init__(
+        self,
+        num_train_timesteps: int = 1024,
+        sigma_data: int = 0.5,
```
Comment (a-r-r-o-w) [inline src/diffusers/schedulers/scheduling_consistency_decoder.py:44]:
```suggestion
        sigma_data: float = 0.5,
```

File: src/diffusers/models/unet_2d.py [hunk 5]
```diff
@@ -144,6 +145,9 @@ def __init__(
         elif time_embedding_type == "positional":
             self.time_proj = Timesteps(block_out_channels[0], flip_sin_to_cos, freq_shift)
             timestep_input_dim = block_out_channels[0]
+        elif time_embedding_type == "learned":
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d.py:148]: Super cool that it works out of the box with our unet!

File: src/diffusers/models/unet_2d.py [hunk 6]
```diff
@@ -144,6 +145,9 @@ def __init__(
...
+            self.time_proj = nn.Embedding(num_train_timesteps, block_out_channels[0])
```
Comment (DN6) [inline src/diffusers/models/unet_2d.py:149]: Should `num_train_timesteps` default to an int value?  What happens if `time_embedding_type` is set to `learned` and no timesteps are provided?

File: src/diffusers/pipelines/consistency_models/pipeline_consistency_decoder.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+                returned where the first element is a list with the generated images.
+        """
+        means = torch.tensor([0.38862467, 0.02253063, 0.07381133, -0.0171294], device=z.device, dtype=z.dtype)[
+            None, :, None, None
+        ]
+        stds = torch.tensor([0.9654121, 1.0440036, 0.76147926, 0.77022034], device=z.device, dtype=z.dtype)[
+            None, :, None, None
+        ]
+        z = (z - means) / stds
+        z = F.interpolate(z, mode="nearest", scale_factor=8)
```
Comment (DN6) [inline src/diffusers/pipelines/consistency_models/pipeline_consistency_decoder.py:110]: Does scale factor affect output quality? Should it be passed in as a param to `__call__`?

File: src/diffusers/pipelines/consistency_models/pipeline_consistency_decoder.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+            for t in self.scheduler.timesteps:
+                model_input = torch.concat([self.scheduler.scale_model_input(x_t, t), z], dim=1)
+                model_output = self.unet(model_input, t).sample[:, :3, :, :]
+                prev_sample = self.scheduler.step(model_output, t, x_t, generator).prev_sample
+                x_t = prev_sample
+                progress_bar.update()
+
+        x_0 = x_t
+
+        image = self.image_processor.postprocess(x_0, output_type=output_type)
```
Comment (DN6) [inline src/diffusers/pipelines/consistency_models/pipeline_consistency_decoder.py:130]: Why not just pass `x_t` to postprocess directly?

**Review by DN6 — COMMENTED** [review]
LGTM 👍🏽  Had some comments on a few small things.

**Comment by patrickvonplaten** [comment]
Very nice job on porting the checkpoint this quickly @williamberman . 

I feel quite strongly that we should allow this decoder to be a drop-in replacement for existing SD15 pipelines. E.g. the following should work:


```diff
+from diffusers import StableDiffusionPipeline, ConsistencyDecoderVae
-from diffusers import StableDiffusionPipeline
import torch

+vae = ConsistencyDecoderVae.from_pretrained("openai/consistency_decoder", torch_dtype=torch.float16)

model_id = "runwayml/stable-diffusion-v1-5"
+pipe = StableDiffusionPipeline.from_pretrained(model_id, torch_dtype=torch.float16, vae=vae)
-pipe = StableDiffusionPipeline.from_pretrained(model_id, torch_dtype=torch.float16)
pipe = pipe.to("cuda")

prompt = "a photo of an astronaut riding a horse on mars"
image = pipe(prompt).images[0]  
    
image.save("astronaut_rides_horse.png")
```

Therefore, I would do the following:
- 1. Instead of creating a `ConsistencyDecoderVae` class that looks as follows:
```py
class ConsistencyDecoderVae(ModelMixin, ConfigMixin):
    
    @register_to_config
    def __init__(self, <all_config_params_of_unet_and_vae_encoder>):
         # same encoder as standard VAE
         self.encoder = Encoder(
            in_channels=in_channels,
            out_channels=latent_channels,
            down_block_types=down_block_types,
            block_out_channels=block_out_channels,
            layers_per_block=layers_per_block,
            act_fn=act_fn,
            norm_num_groups=norm_num_groups,
            double_z=True,
         )

        self.decoder_unet = UNet2DModel(<all_config_params_on_unet>)
        self.decoder_scheduler = ConsistencyDecoderScheduler()

    def forward(
        self,
        sample: torch.FloatTensor,
        sample_posterior: bool = False,
        return_dict: bool = True,
        generator: Optional[torch.Generator] = None,
    ) -> Union[DecoderOutput, torch.FloatTensor]:
        r"""
        Args:
            sample (`torch.FloatTensor`): Input sample.
            sample_posterior (`bool`, *optional*, defaults to `False`):
                Whether to sample from the posterior.
            return_dict (`bool`, *optional*, defaults to `True`):
                Whether or not to return a [`DecoderOutput`] instead of a plain tuple.
        """
        x = sample
        posterior = self.encode(x).latent_dist
        if sample_posterior:
            z = posterior.sample(generator=generator)
        else:
            z = posterior.mode()
        dec = self.decode(z).sample

        if not return_dict:
            return (dec,)

        return DecoderOutput(sample=dec)

    @apply_forward_hook
    def encode(
        self, x: torch.FloatTensor, return_dict: bool = True
    ) -> Union[AutoencoderKLOutput, Tuple[DiagonalGaussianDistribution]]:
        """
        Encode a batch of images into latents.

        Args:
            x (`torch.FloatTensor`): Input batch of images.
            return_dict (`bool`, *optional*, defaults to `True`):
                Whether to return a [`~models.autoencoder_kl.AutoencoderKLOutput`] instead of a plain tuple.

        Returns:
                The latent representations of the encoded images. If `return_dict` is True, a
                [`~models.autoencoder_kl.AutoencoderKLOutput`] is returned, otherwise a plain `tuple` is returned.
        """
        if self.use_tiling and (x.shape[-1] > self.tile_sample_min_size or x.shape[-2] > self.tile_sample_min_size):
            return self.tiled_encode(x, return_dict=return_dict)

        if self.use_slicing and x.shape[0] > 1:
            encoded_slices = [self.encoder(x_slice) for x_slice in x.split(1)]
            h = torch.cat(encoded_slices)
        else:
            h = self.encoder(x)

        moments = self.quant_conv(h)
        posterior = DiagonalGaussianDistribution(moments)

        if not return_dict:
            return (posterior,)

        return AutoencoderKLOutput(latent_dist=posterior)

    @apply_forward_hook
    def decode(self, z: torch.FloatTensor, return_dict: bool = True) -> Union[DecoderOutput, torch.FloatTensor]:
        # here call the scheduler and the unet ...
```


For now let's maybe just not allow saving the and loading the scheduler and always default to the correct default values until we have more usage (we can always add it later IMO)

**Comment by sayakpaul** [comment]
But @patrickvonplaten won't that be restrictive because users won't be able to control the number of timesteps for the decoder scheduler (I know it's currently not supported). 

Do you think most users would stick to 2 timesteps? I think they would.

File: src/diffusers/pipelines/consistency_models/pipeline_consistency_decoder.py (hunk 7 above)
Comment (williamberman) [reply src/diffusers/pipelines/consistency_models/pipeline_consistency_decoder.py:110]: scale factor is required to be the same as the downsampling factor of the encoding

File: src/diffusers/pipelines/consistency_models/pipeline_consistency_decoder.py (hunk 8 above)
Comment (williamberman) [reply src/diffusers/pipelines/consistency_models/pipeline_consistency_decoder.py:130]: the variable rename is for documentation that at the end of the denoising loop we have t==0

**Comment by williamberman** [comment]
yeah chatted with patrick here, I have no strong opinions on pipeline vs model class

File: src/diffusers/models/unet_2d.py (hunk 6 above)
Comment (williamberman) [reply src/diffusers/models/unet_2d.py:149]: I think it's acceptable to assume someone hitting this code path will know to provide the right value, and if not pytorch will throw an error

File: src/diffusers/pipelines/consistency_models/pipeline_consistency_decoder.py (hunk 3 above)
Comment (williamberman) [reply src/diffusers/pipelines/consistency_models/pipeline_consistency_decoder.py]: was able to move to buffers on the vae now that it's a module

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 9]
```diff
@@ -838,7 +838,9 @@ def __call__(
                         callback(step_idx, t, latents)
 
         if not output_type == "latent":
-            image = self.vae.decode(latents / self.vae.config.scaling_factor, return_dict=False)[0]
+            image = self.vae.decode(latents / self.vae.config.scaling_factor, return_dict=False, generator=generator)[
+                0
+            ]
```
Comment (williamberman) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:843]: This helps for tests for the noise generation in the decoder, I also added an un-used generator arg to the AutoencoderKL.decode so should be fine

File: tests/models/test_modeling_common.py [hunk 10]
```diff
@@ -196,11 +196,15 @@ def test_forward_with_norm_groups(self):
...
     base_precision = 1e-3
+    forward_requires_fresh_args = False
 
     def test_from_save_pretrained(self, expected_max_diff=5e-5):
-        init_dict, inputs_dict = self.prepare_init_args_and_inputs_for_common()
+        if self.forward_requires_fresh_args:
+            model = self.model_class(**self.init_dict)
+        else:
+            init_dict, inputs_dict = self.prepare_init_args_and_inputs_for_common()
+            model = self.model_class(**init_dict)
```
Comment (williamberman) [inline tests/models/test_modeling_common.py:206]: Needed to pass generator with same seed for multiple forward calls

**Comment by williamberman** [comment]
failing tests are unrelated

File: src/diffusers/models/consistency_decoder_vae.py [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+    Args:
+        latent_dist (`DiagonalGaussianDistribution`):
+            Encoded outputs of `Encoder` represented as the mean and logvar of `DiagonalGaussianDistribution`.
+            `DiagonalGaussianDistribution` allows for sampling latents from the distribution.
+    """
+
+    latent_dist: "DiagonalGaussianDistribution"
+
+
+class ConsistencyDecoderVae(ModelMixin, ConfigMixin):
```
Comment (sayakpaul) [inline src/diffusers/models/consistency_decoder_vae.py:52]:
```suggestion
class ConsistencyDecoderVAE(ModelMixin, ConfigMixin):
```

File: docs/source/en/api/schedulers/consistency_decoder.md [hunk 12]
```diff
@@ -0,0 +1,11 @@
+# ConsistencyDecoderScheduler
+
+This scheduler is a part of the [`ConsistencyDecoderPipeline`] and was introduced in [DALL-E 3](https://openai.com/dall-e-3). 
+
+The original codebase can be found at [openai/consistency_models](https://github.com/openai/consistency_models).
+
+## ConsistencyDecoderScheduler
+[[autodoc]] ConsistencyDecoderScheduler
```
Comment (sayakpaul) [inline docs/source/en/api/schedulers/consistency_decoder.md:8]:
```suggestion
```

**Review by sayakpaul — APPROVED** [review]

File: src/diffusers/models/autoencoder_kl.py [hunk 13]
```diff
@@ -294,7 +294,9 @@ def _decode(self, z: torch.FloatTensor, return_dict: bool = True) -> Union[Decod
         return DecoderOutput(sample=dec)
 
     @apply_forward_hook
-    def decode(self, z: torch.FloatTensor, return_dict: bool = True) -> Union[DecoderOutput, torch.FloatTensor]:
+    def decode(
+        self, z: torch.FloatTensor, return_dict: bool = True, generator=None
+    ) -> Union[DecoderOutput, torch.FloatTensor]:
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/models/autoencoder_kl.py:299]: Yes perf

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 9 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:843]: Perf!

File: src/diffusers/pipelines/consistency_models/pipeline_consistency_models.py [hunk 14]
```diff
@@ -1,3 +1,17 @@
+# Copyright 2023 OpenAI and The HuggingFace Team. All rights reserved.
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/consistency_models/pipeline_consistency_models.py:1]:
```suggestion
# Copyright 2023 The HuggingFace Team. All rights reserved.
```

File: src/diffusers/__init__.py [hunk 15]
```diff
@@ -127,6 +128,7 @@
     _import_structure["schedulers"].extend(
         [
             "CMStochasticIterativeScheduler",
+            "ConsistencyDecoderScheduler",
```
Comment (patrickvonplaten) [inline src/diffusers/__init__.py:131]:
```suggestion
```

File: src/diffusers/__init__.py [hunk 16]
```diff
@@ -491,6 +494,7 @@
         )
         from .schedulers import (
             CMStochasticIterativeScheduler,
+            ConsistencyDecoderScheduler,
```
Comment (patrickvonplaten) [inline src/diffusers/__init__.py:497]:
```suggestion
```

File: src/diffusers/models/consistency_decoder_vae.py [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+class ConsistencyDecoderVAE(ModelMixin, ConfigMixin):
+    r"""
+    The consistency decoder used with DALL-E 3.
+
+    Examples:
+        ```py
+        >>> import torch
+        >>> from diffusers import DiffusionPipeline, ConsistencyDecoderVAE
+
+        >>> # TODO - is this going to be where the model is uploaded?
```
Comment (patrickvonplaten) [inline src/diffusers/models/consistency_decoder_vae.py:61]:
```suggestion
```

**Comment by patrickvonplaten** [comment]
Some bug fixes here: https://github.com/huggingface/diffusers/pull/5722
